### DISCUSSION TITLE
Hotjar support available in Lit 2.x?

### DISCUSSION THREAD

**Discussion by jeffchew** [lit/lit#2265] [Q&A] [answered]
My team ([Carbon for IBM.com Web Components](https://github.com/carbon-design-system/carbon-for-ibm-dotcom) / [Carbon Web Components](https://github.com/carbon-design-system/carbon-web-components)) is currently on lit-element/lit-html, and currently roadmapping our upgrade to Lit 2.x along with other major version upgrade tasks on our plate.

One of the issues we are noticing is that Hotjar (https://help.hotjar.com/hc/en-us/articles/115012499507-Platforms-and-Frameworks-not-Compatible-with-Hotjar) doesn't appear to work (CSS isn't rendering in heatmap screenshots) and is a tool that our data analyst teams use extensively for analyzing our website traffic.

Anyone know if Hotjar works with Lit 2.x? Or is there something we can do on our end that can get it working on existing lit-element/lit-html?

**Comment by graynorton** [comment, marked as answer]
The long-term solution is for Hotjar to implement support for constructable stylesheets.

Meanwhile, you can force Lit not to use constructable stylesheets by including a script tag like this in the page before loading Lit:

`<script>delete Document.prototype['adoptedStyleSheets'];</script>`

It's kind of gross, but it means we don't need to add configuration complexity to the Lit API and it does exactly what you need to do to make Hotjar work.

**Reply by abdonrd** [reply]
The problem is only with the constructable stylesheets? There are none with the Shadow DOM?

**Reply by graynorton** [reply]
According to the page linked in the original post:

![image](https://user-images.githubusercontent.com/32065/138973327-e56c3d55-6bbb-4337-b837-573311b14b1c.png)

**Reply by jeffchew** [reply]
Yikes, I don't think forcing to disable constructable stylesheets will be an option for us.

I have also reached out to Hotjar support to see when they can add adopted stylesheets support.

But good to know that constructable stylesheets is the underlying reason why it doesn't work today.

**Reply by web-padawan** [reply]
@graynorton would it be an option to reconsider #1811 and add a flag to avoid patching browser API manually?

**Reply by abdonrd** [reply]
Oh, thanks @graynorton! This other section confused me:

<img width="906" alt="Screenshot 2021-10-27 at 10 40 22" src="https://user-images.githubusercontent.com/1007051/139031156-31d2161f-e3e9-454d-a4e1-9dec5bc2421f.png">

It must be an older mention.

Glad to hear that Hotjar already supports Shadow DOM!

**Reply by jeffchew** [reply]
Thank you for the response, I might actually do some testing with removing adopted stylesheets to see if this is a workaround for us for now. And as mentioned, I have already reached out to the Hotjar team to see how quickly they can adopt adopted stylesheets.

**Reply by graynorton** [reply]
@web-padawan We'd prefer to avoid adding Lit API surface and code complexity for this unless it proves absolutely necessary.

For the use case described in this discussion, it seems like it would be a problem for *anyone* to use constructable stylesheets on the page, not just Lit—so disabling globally actually seems like a more suitable fix.

I think what it would take for us to reconsider #1811 would be for someone to come up with a realistic and important use case where it was necessary to disable constructable stylesheets for all Lit elements, but not for anything else on the page.

**Reply by web-padawan** [reply]
Thank you for the answer. I don't have a strong opinion regarding Lit API shape. You can also close #1811 as won't fix.
